File: plugins/typescript-kit/templates/env-schema.ts
/**
 * Named env boundary template (copy into product src/server/env.ts or similar).
 * Prefer a schema lib (zod / valibot / arktype) when the product already has one.
 * Whole-object `process.env` into the parser is allowed; bare `process.env.X`
 * accessors elsewhere still need `ts-rg-allow` on a named helper.
 */
export type AppEnv = {
  NODE_ENV: 'development' | 'test' | 'production'
  PORT: number
  DATABASE_URL?: string
}

type EnvMap = Record<string, string | undefined>

function requireString(raw: string | undefined, key: string): string {
  if (raw === undefined || raw === '') throw new Error(`missing env ${key}`)
  return raw
}

export function parseEnv(raw: EnvMap = process.env): AppEnv {
  const nodeEnv = raw.NODE_ENV ?? 'development'
  if (nodeEnv !== 'development' && nodeEnv !== 'test' && nodeEnv !== 'production') {
    throw new Error('invalid NODE_ENV')
  }
  const portRaw = raw.PORT
  const PORT = portRaw === undefined || portRaw === '' ? 3000 : Number(portRaw)
  if (!Number.isFinite(PORT)) throw new Error('invalid PORT')
  const DATABASE_URL = raw.DATABASE_URL
  return {
    NODE_ENV: nodeEnv,
    PORT,
    ...(DATABASE_URL !== undefined && DATABASE_URL !== ''
      ? { DATABASE_URL: requireString(DATABASE_URL, 'DATABASE_URL') }
      : {}),
  }
}

export const env = parseEnv()

File: plugins/python-kit/templates/env_schema.py
"""Named env boundary template (copy into product src/env_schema.py or similar).

Prefer pydantic / attrs / dataclasses when the product already has one.
Whole-dict `os.environ` into the parser is the intended pattern; bare
`os.environ[KEY]` elsewhere still needs `py-rg-allow` on a named helper.
"""

from __future__ import annotations

import os
from typing import Mapping


class AppEnv(dict[str, str]):
    pass


def parse_env(raw: Mapping[str, str] | None = None) -> AppEnv:
    data = dict(raw) if raw is not None else dict(os.environ)  # py-rg-allow named env boundary
    node_env = data.get("NODE_ENV") or data.get("FLASK_ENV") or "development"
    if node_env not in {"development", "test", "production"}:
        raise ValueError("invalid NODE_ENV/FLASK_ENV")
    port_raw = data.get("PORT", "3000")
    try:
        port = int(port_raw)
    except ValueError as exc:
        raise ValueError("invalid PORT") from exc
    out = AppEnv(NODE_ENV=node_env, PORT=str(port))
    if database_url := data.get("DATABASE_URL"):
        out["DATABASE_URL"] = database_url
    return out


env = parse_env()
